apps/aecore/lib/aecore/contract/contract.ex

```diff
@@ -0,0 +1,167 @@
+defmodule Aecore.Contract.Contract do
+  @moduledoc """
+  Aecore contact module implementation.
+  """
+  alias Aecore.Contract.Contract
+  alias Aecore.Chain.Identifier
+  alias Aeutil.Hash
+
+  @version 1
+
+  # The @store_prefix is used to name the storage tree and keep
+  # all storage nodes in one subtree under the contract tree.
+  @store_prefix 16
+
+  @type contract :: %Contract{
+          id: Identifier.t(),
+          owner: Identifier.t(),
+          vm_version: byte(),
+          code: binary(),
+          store: %{binary() => binary()},
+          log: binary(),
+          active: boolean(),
+          referers: [Identifier.t()],
+          deposit: non_neg_integer()
+        }
+
+  @type t :: contract()
+
+  defstruct [:id, :owner, :vm_version, :code, :store, :log, :active, :referers, :deposit]
+
+  use Aecore.Util.Serializable
+
+  @spec new(Keys.pubkey(), non_neg_integer(), byte(), binary(), non_neg_integer()) :: contract()
+  def new(owner, nonce, vm_version, code, deposit) do
+    contract_id = create_contract_id(owner, nonce)
+    identified_contract = Identifier.create_identity(contract_id, :contract)
+    identified_owner = Identifier.create_identity(owner, :account)
+
+    %Contract{
+      id: identified_contract,
+      owner: identified_owner,
+      vm_version: vm_version,
+      code: code,
+      store: %{},
+      log: <<>>,
+      active: true,
+      referers: [],
+      deposit: deposit
+    }
+  end
+
+  @spec encode_to_list(Contract.t()) :: list()
+  def encode_to_list(%Contract{
+        owner: owner,
+        vm_version: vm_version,
+        code: code,
+        log: log,
+        active: active,
+        referers: referers,
+        deposit: deposit
+      }) do
+    encoded_active =
+      case active do
+        true -> 1
+        false -> 0
+      end
+
+    encoded_referers =
+      Enum.map(referers, fn referer ->
+        Identifier.encode_to_binary(referer)
+      end)
+
+    [
+      @version,
+      Identifier.encode_to_binary(owner),
+      :binary.encode_unsigned(vm_version),
+      code,
+      log,
+      encoded_active,
+      encoded_referers,
+      :binary.encode_unsigned(deposit)
+    ]
+  end
+
+  @spec decode_from_list(integer(), list()) :: {:ok, t()} | {:error, String.t()}
+  def decode_from_list(@version, [
+        owner,
+        vm_version,
+        code,
+        log,
+        active,
+        referers,
+        deposit
+      ]) do
+    decoded_active = decode_active(active)
+
+    decoded_referers =
+      Enum.reduce_while(referers, [], fn referer, acc ->
+        with {:ok, decoded_referer} <- Identifier.decode_from_binary(referer) do
+          [decoded_referer | acc]
```

Comment:
This possibly will crash. 
`Enum.reduce_while` requires tuple `{:cont, var}` in order to continue. And tuple `{:halt, var}` to exit the reduce (which you already have).
Additionally, the way this is written, in the first time the check fails, you will exit the reduce with the `{error, reason}` value for `decoded_referers`. I don't know if this is the required end result.

The `with` statement here could be changed for a simple `case`, because you do only one check, therefore no need for it in this example.